lib/Optimizer/Transforms/DecompositionPatterns.cpp

```diff
@@ -114,24 +249,29 @@ struct HToPhasedRx : public OpRewritePattern<quake::HOp> {
                                 PatternRewriter &rewriter) const override {
     if (!op.getControls().empty())
       return failure();
-    if (!quake::isAllReferences(op))
+    const bool allWires = isAllWires(op);
+    if (!quake::isAllReferences(op) && !allWires)
```

Comment:
> I wonder if you still need this test. Quantum operation are either memory semantics (all references) or value semantics (all wires). A mix should be ill-formed. If you can handle both, there shouldn't be a need to test.

@schweitzpgi - is this check already part of the verifiers somewhere? If so, I may be able to remove it as @boschmitt is suggesting. (Though I still would want to exclude `!quake.control` types because, as I understand it, those are not linear values.)

Reply:
We need to support intermediate forms and other types (such as `!quake.control`). I would recommend removing the "all this or that" test(s), and just dealing with the arguments in a type correct way.

Reply:
> We need to support intermediate forms and other types (such as `!quake.control`). I would recommend removing the "all this or that" test(s), and just dealing with the arguments in a type correct way.

I believe that means we will have to do `quake.from_ctrl` and `quake.to_ctrl` in cases shown below.

https://github.com/NVIDIA/cuda-quantum/blob/c4001f66a8f90c0367126b0fdc5df59f24cd8cdf/lib/Optimizer/Transforms/DecompositionPatterns.cpp#L487
https://github.com/NVIDIA/cuda-quantum/blob/c4001f66a8f90c0367126b0fdc5df59f24cd8cdf/lib/Optimizer/Transforms/DecompositionPatterns.cpp#L657
(Cases like the ones linked above perform gates _on the controls_, treating the controls _as targets_ for part of the decompositions.)

In other words, just because a qubit is used as a `quake.control` on the _input_ to the decomposition, doesn't mean that it can remain a `quake.control` _throughout_ the decomposition.

Reply:
The decomposition algorithm used in the "CCZToCX" pattern appears to be fundamentally incompatible with the notion of accepting `!quake.control` types because it modifies the state of the control qubit.
https://github.com/NVIDIA/cuda-quantum/blob/c4001f66a8f90c0367126b0fdc5df59f24cd8cdf/lib/Optimizer/Transforms/DecompositionPatterns.cpp#L602-L610

Reply:
> The decomposition algorithm used in the "CCZToCX" pattern appears to be fundamentally incompatible with the notion of accepting `!quake.control` types because it modifies the state of the control qubit.

FYI, this pattern illustrates well one reason why we decided to go back on our initial design in which controls did not generate new values. We concluded it was a bad tradeoff.

One one hand, not generating new values for controls allowed us to easily identify equivalences like:
```
   (1)  (2)  (3)       (2)  (1)  (3)       (1)  (3)  (2)        
────●────●────●───   ───●────●────●───   ───●────●────●───
    │    │    │         │    │    │         │    │    │  
    │  ┌─┴─┐  │       ┌─┴─┐  │    │         │    │  ┌─┴─┐
────┼──┤ x ├──┼─── = ─┤ x ├──┼────┼─── = ───┼────┼──┤ x ├─
    │  └───┘  │       └───┘  │    │         │    │  └───┘
  ┌─┴─┐     ┌─┴─┐          ┌─┴─┐┌─┴─┐     ┌─┴─┐┌─┴─┐     
──┤ x ├─────┤ x ├─   ──────┤ x ├┤ x ├─   ─┤ x ├┤ x ├──────
  └───┘     └───┘          └───┘└───┘     └───┘└───┘     
```
Mitigating the problem of structural bias, and thus enhancing the effectiveness of some optimizations (e.g., identifying that (1) and (3) cancel out).

On the other hand, this idea made changing the IR more error prone in general. For example, this rewriting pattern changes controls to targets, thus all uses of the control value from the rewriting point onwards would need to change, but the preceding uses would need to stay the same.

Reply:
For posterity: These transitions from `!quake.control` to targets can be handled seamlessly with the use of `quake::FromControlOp` and `quake::ToControlOp`. For the above, after the seventh gate (X), we could insert a `quake::FromControlOp` and feed that wire into the eighth gate (X). Similarly for the top most wire in the figure. Finally, we'd add two `quake::ToControlOp` to convert the wire back to a control type for type correctness.

Reply:
Is it safe to assume that all uses a `!quake.control` have to exist in same block that generated it? In other words, is the following IR not allowed?

```mlir
func.func @test(%arg0: !quake.ref, %arg1: !quake.ref, %arg2: !quake.ref) {
  %0 = quake.unwrap %arg0 : (!quake.ref) -> !quake.wire
  %1 = quake.unwrap %arg1 : (!quake.ref) -> !quake.wire
  %2 = quake.unwrap %arg2 : (!quake.ref) -> !quake.wire
  %c0 = quake.to_ctrl %0 : (!quake.wire) -> !quake.control
  %cond = cc.undef i1
  %3:2 = cc.if(%cond) ((%a1 = %1, %a2 = %2)) -> (!quake.wire, !quake.wire) {
    %4:2 = quake.z [%c0, %a1] %a2 : (!quake.control, !quake.wire, !quake.wire) -> (!quake.wire, !quake.wire)
    cc.continue %4#0, %4#1 : !quake.wire, !quake.wire
  } else {
    cc.continue %a1, %a2 : !quake.wire, !quake.wire
  }
  %5 = quake.from_ctrl %c0 : (!quake.control) -> !quake.wire
  quake.wrap %5 to %arg0 : !quake.wire, !quake.ref
  quake.wrap %3#0 to %arg1 : !quake.wire, !quake.ref
  quake.wrap %3#1 to %arg2 : !quake.wire, !quake.ref
  return
}
```
The above IR currently passes all validation checks, but if we performed a decomposition on it, we would have nowhere to pass the new `!quake.control` value generated by the decomposition to (and the downstream code may attempt to keep using `%c0` or `%5` despite that no longer corresponding to the correct value anymore.

I was assuming that the above IR is valid, and it was therefore unsafe to try to do what you are suggesting, but maybe this should be flagged as invalid IR?

Or maybe we should use `isUsedOutsideOfBlock` to detect this condition and only `return failure()` in this condition?

Reply:
This IR is fine.

It's a bit odd in that the region of the control is unnecessarily too large. So one strategy would be to thread wires into the if and move the to_control and from_control within the region they actually make sense.

Another valid transformation for this would have to take into account that `cc.if` has regions and definitions in those regions are not upwards exposed and realize the definition and at least 1 other use are not in the same region. So we could do something like:

```mlir
  %c0 = to_ctrl
  %if:3 = cc.if ...  {
    ... %c0 ....
    ...
    %new_wire = from_ctrl %c0
    ....
    %new_control = to_ctrl %new_wire'
    continue %new_control, ...
  } else {
    continue %c0, ...
  }
  ...
  %wire = from_control %if#0
```

So, yeah, it's still possible, but we have to maintain a valid SSA representation on the control typed values.

Reply:
```mlir
module {
  func.func @test(%arg0: !quake.ref, %arg1: !quake.ref, %arg2: !quake.ref) {
    %0 = quake.unwrap %arg0 : (!quake.ref) -> !quake.wire
    %1 = quake.unwrap %arg1 : (!quake.ref) -> !quake.wire
    %2 = quake.unwrap %arg2 : (!quake.ref) -> !quake.wire
    %3 = quake.to_ctrl %0 : (!quake.wire) -> !quake.control
    %4 = cc.undef i1
    %5:3 = cc.if(%4) ((%arg3 = %1, %arg4 = %2)) -> (!quake.control, !quake.wire, !quake.wire) {
      %7:2 = quake.x [%arg3] %arg4 : (!quake.wire, !quake.wire) -> (!quake.wire, !quake.wire)
      %8 = quake.t<adj> %7#1 : (!quake.wire) -> !quake.wire
      %9 = quake.x [%3] %8 : (!quake.control, !quake.wire) -> !quake.wire
      %10 = quake.t %9 : (!quake.wire) -> !quake.wire
      %11:2 = quake.x [%7#0] %10 : (!quake.wire, !quake.wire) -> (!quake.wire, !quake.wire)
      %12 = quake.t<adj> %11#1 : (!quake.wire) -> !quake.wire
      %13 = quake.x [%3] %12 : (!quake.control, !quake.wire) -> !quake.wire
      %14 = quake.x [%3] %11#0 : (!quake.control, !quake.wire) -> !quake.wire
      %15 = quake.t<adj> %14 : (!quake.wire) -> !quake.wire
      %16 = quake.x [%3] %15 : (!quake.control, !quake.wire) -> !quake.wire
      %17 = quake.t %16 : (!quake.wire) -> !quake.wire
      %18 = quake.t %13 : (!quake.wire) -> !quake.wire
      %19 = quake.from_ctrl %3 : (!quake.control) -> !quake.wire
      %20 = quake.t %19 : (!quake.wire) -> !quake.wire
      %21 = quake.to_ctrl %20 : (!quake.wire) -> !quake.control
      cc.continue %21, %17, %18 : !quake.control, !quake.wire, !quake.wire
    } else {
      cc.continue %3, %arg3, %arg4 : !quake.control, !quake.wire, !quake.wire
    }
    %6 = quake.from_ctrl %5#0 : (!quake.control) -> !quake.wire
    quake.wrap %6 to %arg0 : !quake.wire, !quake.ref
    quake.wrap %5#1 to %arg1 : !quake.wire, !quake.ref
    quake.wrap %5#2 to %arg2 : !quake.wire, !quake.ref
    return
  }
}
```

Reply:
OK, thanks. I think I recall a prior discussion about not threading `quake.control`s through `cc.if`s, but if so, that was probably describing the _current_ state of things and not saying that they _couldn't_ be threaded through them. That makes sense now. Thanks.

Reply:
The code posted above is legal and can be run through `cudaq-opt`.

I think the confusion may be that values of type `!quake.control` are SSA. Values of type `!quake.wire` are SSI. In the context of a `cc.if` this distinction matters.

Reply:
Understood. Thanks.

FWIW, both of those Quake snippets in this thread don't seem to make it through `cudaq-opt -regtomem`. I'm not sure if we should create a separate issue for that, or if that is expected behavior? (The first one fails an `assert` check, and the second one doesn't perform any IR transformation.)